docs/metrics/perplexity.md

```diff
@@ -0,0 +1,142 @@
+current generated token is conditioned on.
+
+In pseudocode, the algorithm works as such:
+
+``` py title="Perplexity Pseudocode"
+total_log_likelihood = 0
+word_count = 0
+
+for example in test_dataset:
+    # The correct next word to be predicted
+    reference_word = example.reference
+    # Predicted output distribution
+    predicted_distribution = model.predict(example.context)
+
+    # Add the log-likelihood of the correct next word predicted by the model.
+    total_log_likelihood += log(predicted_distribution[reference_word])
+    word_count += 1
+
+# Calculate the exponential of the negative log likelihood averaged over the
+#     number of words
+perplexity = e ^ (-total_log_likelihood / word_count)
+```
+
+This implementation is simplified and ignores details such as tokenizing the test dataset.
+
+### Interpretation
+
+Ideally, a lower perplexity is desirable. Since perplexity a relative metric with no upper bound, it is best used
+to compare similar models and configurations on a specific task and dataset to look for improvements in model
+performance.
+
+For context, the perplexity score of GPT-2 calculated on the wikitext dataset is 16.4541. The perplexity of
+LLaMA-2 7B calculated on its diverse test dataset is between 5.0 - 6.0. It is not uncommon to see your language
+model perform between these ranges for text-generation tasks.
+
+### Example
+Assume we have a sentence-completion language model that uses a context-length of 4, and contains a very limited
+vocabulary. Let's calculate the perplexity of the hypothetical model.
+
+| Reference | Input |
+| | |
+| `The curious cat explored the mysterious, overgrown garden`. | `The curious cat explored...` |
+
+??? example "Step 1. Generate output distributions"
+    **Forward Pass 1:** <br>
+    $\text{model(}$[`the`, `curious`, `cat`, `explored`] $)$ -> `{`**`the: 45%`**`, a: 25%, with: 10%, ...}`\*<br>
+    \*<small> In this notation, we indicate that model is generating a probability that a given word will be the
+    next word in this sentence. That is, the model believes there is a 45% chance that the next word will be "the",
+    a 25% chance the next word will be "a", and so on. </small>
+
+    **Forward Pass 2:** <br>
+    $\text{model(}$[`curious`, `cat`, `explored`, `the`] $)$ -> `{`**`mysterious: 20%`**`, dark: 15%, colorful: 10%, ...}`<br>
+
+    **Forward Pass 3:** <br>
+    $\text{model(}$[`cat`, `explored`, `the`, `mysterious`] $)$ -> `{`**`overgrown: 70%`**`, and: 12%, underground: 2%, ...}`<br>
+
+    **Forward Pass 4:** <br>
+    $\text{model(}$[`explored`, `the`, `mysterious`, `overgrown`] $)$ -> `{jungle: 60%, cave: 30%, `**`garden: 5%`**`, ...}`<br>
+
+??? example "Step 2. Calculate Perplexity"
+    Our total log likelihood is:
+
+    $$
+    log(0.45) + log(0.2) + log(0.7) + log(0.05) = -2.502
+    $$
+
+    Thus, our perplexity is given by:
+
+    $$
+    \begin{align*}
+    \text{Perplexity} &= exp{\{ \frac{-2.502}{4} \}} \\
```

Comment:
I expect $\exp(a)$ or $e^a$. Let's replace the curly bracket with a round bracket.

<img width="277" alt="image" src="https://github.com/kolenaIO/kolena/assets/17771952/b9b13786-1609-4218-ad04-69ac1a4ec7c5">